File: Dynamic_Programming/Java/longest_increasing_subsequence.java

/**
 * @file longest_increasing_subsequence.java
 * @description Implementation of Longest Increasing Subsequence algorithm in Java
 * @author Algorithm Collection
 * @version 1.0
 */

import java.util.*;

/**
 * @class longest_increasing_subsequence
 * @description Class containing Longest Increasing Subsequence algorithm
 *              implementation
 */
public class longest_increasing_subsequence {

    /**
     * @function lisDP
     * @description Dynamic programming implementation of LIS
     * @param arr int[] - Input array
     * @returns int - Length of longest increasing subsequence
     */
    public static int lisDP(int[] arr) {
        if (arr.length == 0)
            return 0;

        int n = arr.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1);

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[j] < arr[i]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
        }

        return Arrays.stream(dp).max().orElse(0);
    }

    /**
     * @function lisWithSequence
     * @description LIS that returns the actual subsequence
     * @param arr int[] - Input array
     * @returns List<Integer> - Longest increasing subsequence
     */
    public static List<Integer> lisWithSequence(int[] arr) {
        if (arr.length == 0)
            return new ArrayList<>();

        int n = arr.length;
        int[] dp = new int[n];
        int[] parent = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(parent, -1);

        int maxLength = 1;
        int maxIndex = 0;

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[j] < arr[i] && dp[j] + 1 > dp[i]) {
                    dp[i] = dp[j] + 1;
                    parent[i] = j;
                }
            }
            if (dp[i] > maxLength) {
                maxLength = dp[i];
                maxIndex = i;
            }
        }

        // Reconstruct sequence
        List<Integer> sequence = new ArrayList<>();
        int current = maxIndex;
        while (current != -1) {
            sequence.add(arr[current]);
            current = parent[current];
        }

        Collections.reverse(sequence);
        return sequence;
    }

    /**
     * @function lisBinarySearch
     * @description LIS using binary search optimization
     * @param arr int[] - Input array
     * @returns int - Length of longest increasing subsequence
     */
    public static int lisBinarySearch(int[] arr) {
        if (arr.length == 0)
            return 0;

        List<Integer> tail = new ArrayList<>();
        tail.add(arr[0]);

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > tail.get(tail.size() - 1)) {
                tail.add(arr[i]);
            } else {
                int pos = binarySearch(tail, arr[i]);
                tail.set(pos, arr[i]);
            }
        }

        return tail.size();
    }

    /**
     * @function binarySearch
     * @description Binary search helper for LIS
     * @param tail List<Integer> - Tail array
     * @param key  int - Key to search for
     * @returns int - Position to insert/replace
     */
    private static int binarySearch(List<Integer> tail, int key) {
        int left = 0, right = tail.size() - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (tail.get(mid) < key) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }

    /**
     * @function lisWithSequenceBinarySearch
     * @description LIS with sequence using binary search
     * @param arr int[] - Input array
     * @returns List<Integer> - Longest increasing subsequence
     */
    public static List<Integer> lisWithSequenceBinarySearch(int[] arr) {
        if (arr.length == 0)
            return new ArrayList<>();

        List<Integer> tail = new ArrayList<>();
        List<Integer> indices = new ArrayList<>();
        int[] parent = new int[arr.length];
        Arrays.fill(parent, -1);

        tail.add(arr[0]);
        indices.add(0);

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > tail.get(tail.size() - 1)) {
                tail.add(arr[i]);
                indices.add(i);
                if (indices.size() > 1) {
                    parent[i] = indices.get(indices.size() - 2);
                }
            } else {
                int pos = binarySearch(tail, arr[i]);
                tail.set(pos, arr[i]);
                indices.set(pos, i);
                if (pos > 0) {
                    parent[i] = indices.get(pos - 1);
                }
            }
        }

        // Reconstruct sequence
        List<Integer> sequence = new ArrayList<>();
        int current = indices.get(indices.size() - 1);
        while (current != -1) {
            sequence.add(arr[current]);
            current = parent[current];
        }

        Collections.reverse(sequence);
        return sequence;
    }

    /**
     * @function lisStrictlyIncreasing
     * @description LIS for strictly increasing subsequence
     * @param arr int[] - Input array
     * @returns int - Length of longest strictly increasing subsequence
     */
    public static int lisStrictlyIncreasing(int[] arr) {
        if (arr.length == 0)
            return 0;

        List<Integer> tail = new ArrayList<>();
        tail.add(arr[0]);

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > tail.get(tail.size() - 1)) {
                tail.add(arr[i]);
            } else {
                int pos = binarySearch(tail, arr[i]);
                if (tail.get(pos) != arr[i]) {
                    tail.set(pos, arr[i]);
                }
            }
        }

        return tail.size();
    }

    /**
     * @function lisNonDecreasing
     * @description LIS for non-decreasing subsequence
     * @param arr int[] - Input array
     * @returns int - Length of longest non-decreasing subsequence
     */
    public static int lisNonDecreasing(int[] arr) {
        if (arr.length == 0)
            return 0;

        List<Integer> tail = new ArrayList<>();
        tail.add(arr[0]);

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] >= tail.get(tail.size() - 1)) {
                tail.add(arr[i]);
            } else {
                int pos = binarySearchNonDecreasing(tail, arr[i]);
                tail.set(pos, arr[i]);
            }
        }

        return tail.size();
    }

    /**
     * @function binarySearchNonDecreasing
     * @description Binary search for non-decreasing LIS
     * @param tail List<Integer> - Tail array
     * @param key  int - Key to search for
     * @returns int - Position to insert/replace
     */
    private static int binarySearchNonDecreasing(List<Integer> tail, int key) {
        int left = 0, right = tail.size() - 1;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (tail.get(mid) <= key) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return left;
    }

    /**
     * @function lisWithCount
     * @description LIS that returns count of all possible LIS
     * @param arr int[] - Input array
     * @returns int - Count of longest increasing subsequences
     */
    public static int lisWithCount(int[] arr) {
        if (arr.length == 0)
            return 0;

        int n = arr.length;
        int[] dp = new int[n];
        int[] count = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(count, 1);

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[j] < arr[i]) {
                    if (dp[j] + 1 > dp[i]) {
                        dp[i] = dp[j] + 1;
                        count[i] = count[j];
                    } else if (dp[j] + 1 == dp[i]) {
                        count[i] += count[j];
                    }
                }
            }
        }

        int maxLength = Arrays.stream(dp).max().orElse(0);
        int totalCount = 0;

        for (int i = 0; i < n; i++) {
            if (dp[i] == maxLength) {
                totalCount += count[i];
            }
        }

        return totalCount;
    }

    /**
     * @function createSampleData
     * @description Creates sample data for testing
     * @returns int[][] - Array of test cases
     */
    public static int[][] createSampleData() {
        return new int[][] {
                { 10, 9, 2, 5, 3, 7, 101, 18 },
                { 0, 1, 0, 3, 2, 3 },
                { 7, 7, 7, 7, 7, 7, 7 },
                { 1, 3, 6, 7, 9, 4, 10, 5, 6 },
                { 1, 2, 3, 4, 5 },
                { 5, 4, 3, 2, 1 },
                { 1 },
                {}
        };
    }

    /**
     * @function testLIS
     * @description Test function to demonstrate LIS algorithms
     * @returns void
     */
    public static void testLIS() {
        System.out.println("=== Longest Increasing Subsequence Test ===");

        int[][] testCases = createSampleData();

        for (int i = 0; i < testCases.length; i++) {
            int[] arr = testCases[i];
            System.out.println("Test case " + (i + 1) + ": " + Arrays.toString(arr));

            if (arr.length == 0) {
                System.out.println("Empty array - LIS length: 0");
                System.out.println();
                continue;
            }

            // Test DP
            long startTime = System.nanoTime();
            int dpResult = lisDP(arr);
            long endTime = System.nanoTime();
            System.out.printf("DP: %d (%.2f ms)%n", dpResult, (endTime - startTime) / 1_000_000.0);

            // Test Binary Search
            startTime = System.nanoTime();
            int binaryResult = lisBinarySearch(arr);
            endTime = System.nanoTime();
            System.out.printf("Binary Search: %d (%.2f ms)%n", binaryResult, (endTime - startTime) / 1_000_000.0);

            // Test with sequence
            List<Integer> sequence = lisWithSequence(arr);
            System.out.println("LIS sequence: " + sequence);

            // Test strictly increasing
            int strictResult = lisStrictlyIncreasing(arr);
            System.out.println("Strictly increasing: " + strictResult);

            // Test non-decreasing
            int nonDecreasingResult = lisNonDecreasing(arr);
            System.out.println("Non-decreasing: " + nonDecreasingResult);

            // Test count
            int countResult = lisWithCount(arr);
            System.out.println("Count of LIS: " + countResult);

            System.out.println();
        }
    }

    /**
     * @function performanceTest
     * @description Performance test for LIS algorithms
     * @returns void
     */
    public static void performanceTest() {
        System.out.println("=== Performance Test ===");

        int[] sizes = { 1000, 5000, 10000, 20000, 50000 };

        for (int size : sizes) {
            int[] arr = generateRandomArray(size);

            System.out.println("Array size: " + size);

            // Test DP
            long startTime = System.nanoTime();
            lisDP(arr);
            long endTime = System.nanoTime();
            System.out.printf("  DP: %.2f ms%n", (endTime - startTime) / 1_000_000.0);

            // Test Binary Search
            startTime = System.nanoTime();
            lisBinarySearch(arr);
            endTime = System.nanoTime();
            System.out.printf("  Binary Search: %.2f ms%n", (endTime - startTime) / 1_000_000.0);

            System.out.println();
        }
    }

    /**
     * @function generateRandomArray
     * @description Generates a random array of given size
     * @param size int - Size of array to generate
     * @returns int[] - Random array
     */
    private static int[] generateRandomArray(int size) {
        Random random = new Random();
        int[] arr = new int[size];

        for (int i = 0; i < size; i++) {
            arr[i] = random.nextInt(1000);
        }

        return arr;
    }

    /**
     * @function main
     * @description Main method to run the LIS demonstration
     * @param args String[] - Command line arguments
     * @returns void
     */
    public static void main(String[] args) {
        System.out.println("Longest Increasing Subsequence Algorithm Implementation");
        System.out.println("=====================================================");

        // Run basic tests
        testLIS();

        // Run performance test
        performanceTest();

        System.out.println("\nLongest Increasing Subsequence completed successfully!");
    }
}

File: Dynamic_Programming/JavaScript/longest_increasing_subsequence.js
/**
 * @file longest_increasing_subsequence.js
 * @description Implementation of Longest Increasing Subsequence algorithm in JavaScript
 * @author Algorithm Collection
 * @version 1.0
 */

/**
 * @function lisDP
 * @description Dynamic programming implementation of LIS
 * @param {Array<number>} arr - Input array
 * @returns {number} - Length of longest increasing subsequence
 */
function lisDP(arr) {
    if (arr.length === 0) return 0;
    
    const n = arr.length;
    const dp = new Array(n).fill(1);
    
    for (let i = 1; i < n; i++) {
        for (let j = 0; j < i; j++) {
            if (arr[j] < arr[i]) {
                dp[i] = Math.max(dp[i], dp[j] + 1);
            }
        }
    }
    
    return Math.max(...dp);
}

/**
 * @function lisWithSequence
 * @description LIS that returns the actual subsequence
 * @param {Array<number>} arr - Input array
 * @returns {Array<number>} - Longest increasing subsequence
 */
function lisWithSequence(arr) {
    if (arr.length === 0) return [];
    
    const n = arr.length;
    const dp = new Array(n).fill(1);
    const parent = new Array(n).fill(-1);
    
    let maxLength = 1;
    let maxIndex = 0;
    
    for (let i = 1; i < n; i++) {
        for (let j = 0; j < i; j++) {
            if (arr[j] < arr[i] && dp[j] + 1 > dp[i]) {
                dp[i] = dp[j] + 1;
                parent[i] = j;
            }
        }
        if (dp[i] > maxLength) {
            maxLength = dp[i];
            maxIndex = i;
        }
    }
    
    // Reconstruct sequence
    const sequence = [];
    let current = maxIndex;
    while (current !== -1) {
        sequence.push(arr[current]);
        current = parent[current];
    }
    
    return sequence.reverse();
}

/**
 * @function lisBinarySearch
 * @description LIS using binary search optimization
 * @param {Array<number>} arr - Input array
 * @returns {number} - Length of longest increasing subsequence
 */
function lisBinarySearch(arr) {
    if (arr.length === 0) return 0;
    
    const tail = [arr[0]];
    
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] > tail[tail.length - 1]) {
            tail.push(arr[i]);
        } else {
            const pos = binarySearch(tail, arr[i]);
            tail[pos] = arr[i];
        }
    }
    
    return tail.length;
}

/**
 * @function binarySearch
 * @description Binary search helper for LIS
 * @param {Array<number>} tail - Tail array
 * @param {number} key - Key to search for
 * @returns {number} - Position to insert/replace
 */
function binarySearch(tail, key) {
    let left = 0, right = tail.length - 1;
    
    while (left < right) {
        const mid = left + Math.floor((right - left) / 2);
        if (tail[mid] < key) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    
    return left;
}

/**
 * @function lisWithSequenceBinarySearch
 * @description LIS with sequence using binary search
 * @param {Array<number>} arr - Input array
 * @returns {Array<number>} - Longest increasing subsequence
 */
function lisWithSequenceBinarySearch(arr) {
    if (arr.length === 0) return [];
    
    const tail = [arr[0]];
    const indices = [0];
    const parent = new Array(arr.length).fill(-1);
    
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] > tail[tail.length - 1]) {
            tail.push(arr[i]);
            indices.push(i);
            if (indices.length > 1) {
                parent[i] = indices[indices.length - 2];
            }
        } else {
            const pos = binarySearch(tail, arr[i]);
            tail[pos] = arr[i];
            indices[pos] = i;
            if (pos > 0) {
                parent[i] = indices[pos - 1];
            }
        }
    }
    
    // Reconstruct sequence
    const sequence = [];
    let current = indices[indices.length - 1];
    while (current !== -1) {
        sequence.push(arr[current]);
        current = parent[current];
    }
    
    return sequence.reverse();
}

/**
 * @function lisStrictlyIncreasing
 * @description LIS for strictly increasing subsequence
 * @param {Array<number>} arr - Input array
 * @returns {number} - Length of longest strictly increasing subsequence
 */
function lisStrictlyIncreasing(arr) {
    if (arr.length === 0) return 0;
    
    const tail = [arr[0]];
    
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] > tail[tail.length - 1]) {
            tail.push(arr[i]);
        } else {
            const pos = binarySearch(tail, arr[i]);
            if (tail[pos] !== arr[i]) {
                tail[pos] = arr[i];
            }
        }
    }
    
    return tail.length;
}

/**
 * @function lisNonDecreasing
 * @description LIS for non-decreasing subsequence
 * @param {Array<number>} arr - Input array
 * @returns {number} - Length of longest non-decreasing subsequence
 */
function lisNonDecreasing(arr) {
    if (arr.length === 0) return 0;
    
    const tail = [arr[0]];
    
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] >= tail[tail.length - 1]) {
            tail.push(arr[i]);
        } else {
            const pos = binarySearchNonDecreasing(tail, arr[i]);
            tail[pos] = arr[i];
        }
    }
    
    return tail.length;
}

/**
 * @function binarySearchNonDecreasing
 * @description Binary search for non-decreasing LIS
 * @param {Array<number>} tail - Tail array
 * @param {number} key - Key to search for
 * @returns {number} - Position to insert/replace
 */
function binarySearchNonDecreasing(tail, key) {
    let left = 0, right = tail.length - 1;
    
    while (left < right) {
        const mid = left + Math.floor((right - left) / 2);
        if (tail[mid] <= key) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    
    return left;
}

/**
 * @function lisWithCount
 * @description LIS that returns count of all possible LIS
 * @param {Array<number>} arr - Input array
 * @returns {number} - Count of longest increasing subsequences
 */
function lisWithCount(arr) {
    if (arr.length === 0) return 0;
    
    const n = arr.length;
    const dp = new Array(n).fill(1);
    const count = new Array(n).fill(1);
    
    for (let i = 1; i < n; i++) {
        for (let j = 0; j < i; j++) {
            if (arr[j] < arr[i]) {
                if (dp[j] + 1 > dp[i]) {
                    dp[i] = dp[j] + 1;
                    count[i] = count[j];
                } else if (dp[j] + 1 === dp[i]) {
                    count[i] += count[j];
                }
            }
        }
    }
    
    const maxLength = Math.max(...dp);
    let totalCount = 0;
    
    for (let i = 0; i < n; i++) {
        if (dp[i] === maxLength) {
            totalCount += count[i];
        }
    }
    
    return totalCount;
}

/**
 * @function createSampleData
 * @description Creates sample data for testing
 * @returns {Array<Array<number>>} - Array of test cases
 */
function createSampleData() {
    return [
        [10, 9, 2, 5, 3, 7, 101, 18],
        [0, 1, 0, 3, 2, 3],
        [7, 7, 7, 7, 7, 7, 7],
        [1, 3, 6, 7, 9, 4, 10, 5, 6],
        [1, 2, 3, 4, 5],
        [5, 4, 3, 2, 1],
        [1],
        []
    ];
}

/**
 * @function testLIS
 * @description Test function to demonstrate LIS algorithms
 * @returns {void}
 */
function testLIS() {
    console.log("=== Longest Increasing Subsequence Test ===");
    
    const testCases = createSampleData();
    
    for (let i = 0; i < testCases.length; i++) {
        const arr = testCases[i];
        console.log(`Test case ${i + 1}: [${arr.join(', ')}]`);
        
        if (arr.length === 0) {
            console.log("Empty array - LIS length: 0");
            console.log();
            continue;
        }
        
        // Test DP
        const startTime = performance.now();
        const dpResult = lisDP(arr);
        const endTime = performance.now();
        console.log(`DP: ${dpResult} (${(endTime - startTime).toFixed(2)} ms)`);
        
        // Test Binary Search
        const startTime2 = performance.now();
        const binaryResult = lisBinarySearch(arr);
        const endTime2 = performance.now();
        console.log(`Binary Search: ${binaryResult} (${(endTime2 - startTime2).toFixed(2)} ms)`);
        
        // Test with sequence
        const sequence = lisWithSequence(arr);
        console.log(`LIS sequence: [${sequence.join(', ')}]`);
        
        // Test strictly increasing
        const strictResult = lisStrictlyIncreasing(arr);
        console.log(`Strictly increasing: ${strictResult}`);
        
        // Test non-decreasing
        const nonDecreasingResult = lisNonDecreasing(arr);
        console.log(`Non-decreasing: ${nonDecreasingResult}`);
        
        // Test count
        const countResult = lisWithCount(arr);
        console.log(`Count of LIS: ${countResult}`);
        
        console.log();
    }
}

/**
 * @function performanceTest
 * @description Performance test for LIS algorithms
 * @returns {void}
 */
function performanceTest() {
    console.log("=== Performance Test ===");
    
    const sizes = [1000, 5000, 10000, 20000, 50000];
    
    for (const size of sizes) {
        const arr = generateRandomArray(size);
        
        console.log(`Array size: ${size}`);
        
        // Test DP
        const startTime = performance.now();
        lisDP(arr);
        const endTime = performance.now();
        console.log(`  DP: ${(endTime - startTime).toFixed(2)} ms`);
        
        // Test Binary Search
        const startTime2 = performance.now();
        lisBinarySearch(arr);
        const endTime2 = performance.now();
        console.log(`  Binary Search: ${(endTime2 - startTime2).toFixed(2)} ms`);
        
        console.log();
    }
}

/**
 * @function generateRandomArray
 * @description Generates a random array of given size
 * @param {number} size - Size of array to generate
 * @returns {Array<number>} - Random array
 */
function generateRandomArray(size) {
    const arr = [];
    
    for (let i = 0; i < size; i++) {
        arr.push(Math.floor(Math.random() * 1000));
    }
    
    return arr;
}

/**
 * @function main
 * @description Main function to run the LIS demonstration
 * @returns {void}
 */
function main() {
    console.log("Longest Increasing Subsequence Algorithm Implementation");
    console.log("=====================================================");
    
    // Run basic tests
    testLIS();
    
    // Run performance test
    performanceTest();
    
    console.log("\nLongest Increasing Subsequence completed successfully!");
}

// Export functions for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        lisDP,
        lisWithSequence,
        lisBinarySearch,
        lisWithSequenceBinarySearch,
        lisStrictlyIncreasing,
        lisNonDecreasing,
        lisWithCount,
        createSampleData,
        testLIS,
        performanceTest,
        main
    };
}

// Run the main function if this file is executed directly
if (typeof require !== 'undefined' && require.main === module) {
    main();
}
